src/Mod/Draft/draftguitools/gui_edit.py

```diff
@@ -1029,76 +1002,61 @@ def update(self, obj, nodeIndex, v):
-
-        elif objectType == "Sketch":
-            edit_sketcher.updateSketch(obj, nodeIndex, v)
-
-        elif objectType == "Wall":
-            if nodeIndex == 0:
-                edit_arch.updateWall(obj, nodeIndex, v)
-            elif nodeIndex > 0:
-                if obj.Base:
-                    if utils.get_type(obj.Base) in ["Wire", "Circle", "Rectangle",
-                                                    "Polygon", "Sketch"]:
-                        self.update(obj.Base, nodeIndex - 1, v)
-
-        elif objectType == "Window":
-            edit_arch.updateWindow(obj, nodeIndex, v)
-
-        elif objectType == "Space":
-            edit_arch.updateSpace(obj, nodeIndex, v)
-
-        elif objectType == "Structure":
-            edit_arch.updateStructure(obj, nodeIndex, v)
-
-        elif objectType == "PanelCut":
-            edit_arch.updatePanelCut(obj, nodeIndex, v)
+        """Update the object according to the given modified editpoint.
+        If you want to extend the Draft_Edit support you have 2 choices:
+        - provide a .update_object_from_edit_points(obj, nodeIndex, v, self.alt_edit_mode) method 
+          inside your object (receive the object, the node index, the vector and the optional edit mode)
+        - extend the if with your own object type inside Draft_Edit.
+        """
+        obj_gui_tools = self.get_obj_gui_tools(obj)
+        if obj_gui_tools:
+            eps = obj_gui_tools.update_object_from_edit_points(obj, nodeIndex, v, self.alt_edit_mode)
 
-        elif objectType == "PanelSheet":
-            edit_arch.updatePanelSheet(obj, nodeIndex, v)
+        obj.recompute()
 
-        elif objectType == "Part::Line" and obj.TypeId == "Part::Line":
-            edit_part.updatePartLine(obj, nodeIndex, v)
 
-        elif objectType == "Part" and obj.TypeId == "Part::Box":
-            edit_part.updatePartBox(obj, nodeIndex, v)
+    # -------------------------------------------------------------------------
+    # UTILS
+    # -------------------------------------------------------------------------
 
-        elif objectType == "Part" and obj.TypeId == "Part::Cylinder":
-            edit_part.updatePartCylinder(obj, nodeIndex, v)
+    def has_obj_gui_tools(self, obj):
+        """ Check if the object has the GuiTools to provide information to edit it.
+        """
+        if (hasattr(obj, 'obj_gui_tools') or
+            (hasattr(obj, 'Proxy') and hasattr(obj.Proxy, 'obj_gui_tools')) or
+            (self.get_obj_type(obj) in self.gui_tools_repository.keys()) ):
+            return True
+        else:
+            return False
 
-        elif objectType == "Part" and obj.TypeId == "Part::Cone":
-            edit_part.updatePartCone(obj, nodeIndex, v)
 
-        elif objectType == "Part" and obj.TypeId == "Part::Sphere":
-            edit_part.updatePartSphere(obj, nodeIndex, v)
+    def get_obj_type(self, obj):
+        """ Substitute utils.get_type to provide more specific object support.
```

Comment:
I don't understand this change. Maybe `utils.get_type` should be augmented?

As it is right now, it seems to me that this method just wraps around the `get_type` function.
```py
def get_obj_type(self, obj):
    return utils.get_type(obj)
```

Maybe there is an idea that you want to implement that I am not seeing at the moment.

Reply:
Yes, the point is that utils.get_tipe is really limited. I need the function to return the obj.Proxy.Type or the obj.TypeId if the object do not have a Proxy.
So we can know if a Part object is a Box or a Cylinder.

Reply:
Mmm. I think we can. Draft does not test many Part objects anyway, these are special cases; and finding these special cases in the code shouldn't take a long time. Draft and Arch mostly test for Draft and Arch objects, which have a `Proxy.Type`. For other cases we could just return the `TypeId`, just like you do here.
```py
        if hasattr(obj, 'TypeId'):
            return obj.TypeId
```

Reply:
Ok, perfect so! I'll change get_type.

Reply:
You'd need to adjust the lines that use `get_type` with the deleted types. I did a quick search in Draft and Arch and this is what I found. These are the types that don't use the `Proxy.Type` nor the base `TypeId`.
* `"Shape"`, used in `make_sketch`
* `"Sketch"`, used in `gui_edit`, and `ArchWall`
* `"Part::Line"`, used in `gui_trimex`, and `gui_edit`
* `"Offset2D"`, not used anywhere
* `"Part"`, used in `gui_edit`, `offset`, and `importIFClegacy`
* `"Annotation"`, used in `importDXF`, `getSVG`, and `move`
* `"Mesh"`, used in `gui_snapper`
* `"Points"`, used in `gui_snapper`
* `"Group"`, used in `importIFClegacy`

So, it doesn't seem that a lot of code needs to be changed. You could also make a small change to keep compatibility.
```py
if hasattr(obj, 'TypeId'):
    if obj.TypeId == "Part::Offset2D":
        return "Offset2D"
    elif obj.TypeId == "App::Annotation":
        return "Annotation"
    ...
    return obj.TypeId
```

Reply:
ok, let's do it:

- [ ]     "Shape", used in make_sketch
- [ ]     "Sketch", used in gui_edit, and ArchWall
- [ ]     "Part::Line", used in gui_trimex, and gui_edit
- [ ]     "Offset2D", not used anywhere
- [ ]     "Part", used in gui_edit, offset, and importIFClegacy
- [ ]     "Annotation", used in importDXF, getSVG, and move
- [ ]     "Mesh", used in gui_snapper
- [ ]     "Points", used in gui_snapper
- [ ]     "Group", used in importIFClegacy

Reply:
@vocx-fc Is there a rationale for using `TypeId` rather than `isinstance`? Is it because of C++ bridge or something else?

Reply:
Well, `TypeId` is a native property that objects have, so I think we can just test for it. With `isinstance` we would need to load the class type to test for it. What is the advantage?

Reply:
@vocx-fc The advantage of `isinstance` is twofold:

- it's a part of Python language and provides strict guarantees. While TypeId is a weak string which is a subject to human error.
- isinstance is supported by the type inference tools:

![Screenshot from 2020-08-17 05-58-03](https://user-images.githubusercontent.com/361985/90353664-78167f80-e04f-11ea-8406-adec84d9c2e5.png)

This not only helps developers who use IDE (proper autocomplete support), but also guarantees that developers who don't use IDE make no mistake of accessing non-existing property of an object (e.g. mypy linter).

The type of issue that can be prevented by this is demonstrated here:
https://forum.freecadweb.org/viewtopic.php?p=424088#p424088

Reply:
Well, in this particular case, not every single object has a specific type. For example, a `Part::Cylinder` is an instance of type `Part.Feature`, but it is *not* an instance of `Part.Cylinder`.

Most classes under `Part` I think are wrappers of the OpenCASCADE types. There is a difference between the class of a "shape", and the class of an "object". Objects are what you actually see in the tree view, but I'm not sure how the internal implementations are made. You'd need to ask Werner.

`Part.Line` is not an object, it is a class that determines a topological shape of a line (`Part::TopoShape`), while `obj.TypeId == 'Part::Line'` tells us this is a specific object; it is a specific `Part::Feature` object with certain properties that makes it behave like a parametric line.
```py
>>> isinstance(App.ActiveDocument.Line, Part.Feature)
True
>>> isinstance(App.ActiveDocument.Line, Part.Line)
False
```

And about loading `Part` anyway, yes, my understanding is that we require loading `Part` in many cases, however, Yorik, the developer of Draft/Arch, has always said that we should avoid loading `Part` unless we really need it because otherwise it makes loading of the workbench slow. So, we should delay as much as possible the use of `Part` classes, unless we really need them. Therefore, to test a line, we can simply test the `obj.TypeId`, and often this is enough, instead of loading and testing the type itself (`Part.Line` or something else).

> Does it have something to do with serialization?

I don't think it does. Maybe it's due to the way the wrapping of C++ classes is done. I'm not entirely sure how this works, it is a mystery to me. You'd have to ask Werner.

Reply:
Ok, I see now. FreeCAD has its own `DocumentObject` type system defined using the `PROPERTY_SOURCE` macro. And properties of `DocumentObject` are defined using the `ADD_PROPERTY_TYPE` macro.

For example, `Part::Line` is defined in `PrimitiveFeature.cpp`, `Part::Primitive` class:

https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Mod/Part/App/PrimitiveFeature.cpp#L250-L260

The macro calls this static method to associate the `Part::Line` string with the `Part::Primitive` C++ class:
https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Base/Type.cpp#L142-L153

And then, when you use `Document::addObject`, it calls the instantiation method:

https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Base/Type.cpp#L91-L103

This implementation means, that if you have an object with `TypeId` that starts with `Modulename::`, `Modulename` has already been imported. Otherwise you wouldn't have the object created in the first place.

Ok. This appears to have everything to do with serialization: to be able to change C++ implementation while keeping the FCStd file's object type names.

So, when you check `if obj.TypeId == 'Part::Line'`, the only thing that is guaranteed that this object has properties `X1, Y1, Z1, X2, Y2, Z2`. It means you still have to use `isinstance` to call the methods of this object. The Python class corresponding to `Part::Line` is `PrimitivePy`. The inheritance chain:

`PrimitivePy > PartFeaturePy > App::GeoFeaturePy > DocumentObjectPy > App::ExtensionContainerPy > App::PropertyContainerPy > Base::PersistencePy > Base::BaseClassPy > Base::PyObjectBase > PyObject`

To sum up:
- if we want to access object data, we may use `TypeId`
- if we want to call object methods, we must use `isinstance`
- not all *Py types are exported, so isinstance can't always be used. This can be fixed
- DocumentObject properties defined with `ADD_PROPERTY_TYPE` are not part of *Py types. This can be fixed as well

Reply:
Yes, for low level details like these, please ask Werner. I've never tried to understand all the C++ code because for user facing tools it's not very important; the current interfaces that are accessible from Python are enough to create user tools, so I've been content with improving these tools, as well as organizing the Python code so that it is easier to extend by others.

If you think it's helpful, I also recommend you to document your findings in the wiki so that other C++ programmers can get into the code faster. Two resources you should be aware of:
* [FreeCAD_Mod_Dev_Guide](https://github.com/qingfengxia/FreeCAD_Mod_Dev_Guide), written by Qingfeng, the original author of the Cfd workbench. This guide describes some details on how the core is structured. He started writing it for v0.16 and I believe he has updated some parts for v0.19. He intentionally placed it in GitHub so that other people could contribute, but so far I don't think anybody but him has added content.
* Realthunder's [Assembly3 repository](https://github.com/realthunder/FreeCAD_assembly3/wiki). He has added a lot of information of his changes with respect to the main branch. Many of these changes were already merged August 2019 with the `LinkMerge` branch, and possibly more commits will come as soon as 0.19 is released as stable.